### PR TITLE
pkgconf: drop-in replacement for `pkg-config`

### PR DESCRIPTION
**Pull request by cho-m**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Review by carlocab — APPROVED** [review]
I suggest retiring `pkg-config` already too. i.e. Let's delete it and add a formula rename for `pkg-config -> pkgconf`.

**Comment by cho-m** [comment]
> I suggest retiring `pkg-config` already too. i.e. Let's delete it and add a formula rename for `pkg-config -> pkgconf`.

That's a bit complicated as audit doesn't like usage of old names / aliases.

Assuming audit is not changed, either need to:
* Do a giant migration of all `pkg-config` usage to `pkgconf`.
* Make `pkg-config` into a transitional formula that just installs `pkgconf` and then do the migration at a more leisurely pace.

**Comment by carlocab** [comment]
We can just add a temporary exception for `pkg-config` and do the migration gradually.

**Comment by Bo98** [comment]
Am OK with this PR given `pkgconf` and `pkg-config` are already conflicting.

In terms of switching our uses to `pkgconf`, the requirements are:

* `pkgconf` cannot conflict with `pkg-config`
* `brew install pkg-config` must continue to work (I'd probably go as far as saying it should continue to do so without warnings, i.e. as an alias)

Merging via renames and merging via alias however works completely differently. One uses the migrator mechanism and one doesn't. So will need to carefully test it.

I expect anything we do here to be "special", which may indeed adjusting things in `brew` to allow `depends_on "pkg-config"` (potentially for a long time/forever as it would affect so many third-party formulae).

File: Formula/p/pkgconf.rb [hunk 1]
```diff
@@ -4,6 +4,7 @@ class Pkgconf < Formula
   url "https://distfiles.ariadne.space/pkgconf/pkgconf-2.3.0.tar.xz"
```
Comment (cho-m) [inline Formula/p/pkgconf.rb:4]: Need to see if anyone has mirror. Maybe also switch to use `.gz`?

**Comment by cho-m** [comment]
> Merging via renames and merging via alias however works completely differently. One uses the migrator mechanism and one doesn't. So will need to carefully test it.

At least `pkgconf` is using a higher version number so can be seen as an upgrade.
```console
❯ brew outdated pkg-config
pkgconf (0.29.2_3) < 2.3.0_1
```

Though that also means the old version may remain for users who avoid upgrades.

**Comment by cho-m** [comment]
For alias option, probably need https://github.com/Homebrew/brew/pull/18675 in stable tag before continuing.

**Comment by cho-m** [comment]
I guess one possible issue is if we update shim as a user that still has old `pkg-config` may not end up with `pkgconf` path:

https://github.com/Homebrew/brew/blob/master/Library/Homebrew/shims/mac/super/pkg-config#L5

---

Though, perhaps we shouldn't be using the explicit path. It does cause `pkg-config` to be silently added as a dependency as long as it is locally installed.

Maybe using the `PATH` lookup would align closer to our other shims.

**Comment by cho-m** [comment]
Alias support available in 4.4.4. I think `brew` should just see this as an upgrade and handle like other version bumps.

Only bug I see is `brew` may not clean up the old Cellar.

**Comment by carlocab** [comment]
LGTM. CC @Homebrew/core for additional feedback.

**Comment by MikeMcQuaid** [comment]
> Let's delete it and add a formula rename for `pkg-config -> pkgconf`.

I don't think we should do this. It's not a formula rename, it's another formula that we're migrating to.

I feel vaguely uneasy but OK about the `pkg-config` alias but the above is definitely a step too far.

> Alias support available in 4.4.4. I think `brew` should just see this as an upgrade and handle like other version bumps.

@cho-m can you make sure you've tested this locally? Want to ensure that this is robust before this is merged.

**Review by MikeMcQuaid — APPROVED** [review]
Fine with me when @cho-m is 👍🏻 and has tested upgrade path locally.

**Comment by cho-m** [comment]
At least the basic usage is all fine from tap. Testing without revision on `pkgconf` to check bottle usage.

* `brew upgrade` will install `pkgconf` and link it, e.g.
  ```console
  $ brew outdated
  pkgconf (0.29.2_3) < 2.3.0
  
  $ brew upgrade
  ==> Upgrading 1 outdated package:
  pkg-config 2.3.0
  ==> Downloading https://ghcr.io/v2/homebrew/core/pkgconf/manifests/2.3.0
  Already downloaded: /Users/cho-m/Library/Caches/Homebrew/downloads/2f1a3719bf003aa9443a2e2e6af8e6441aa287088a24771231f191a9bcf36884--pkgconf-2.3.0.bottle_manifest.json
  ==> Fetching pkgconf
  ==> Downloading https://ghcr.io/v2/homebrew/core/pkgconf/blobs/sha256:9b44fe313d296fa10617d6b58ea1eab78bf217b6f00478ad98b01e0375475472
  Already downloaded: /Users/cho-m/Library/Caches/Homebrew/downloads/eed8a0a3462be8896edfce1cf6d726e4ebfd489cafecfb485f1b1b669a5d89af--pkgconf--2.3.0.sequoia.bottle.tar.gz
  ==> Upgrading pkg-config
    -> 2.3.0
  ==> Pouring pkgconf--2.3.0.sequoia.bottle.tar.gz
  🍺  /usr/local/Cellar/pkgconf/2.3.0: 25 files, 327.9KB
  ==> Running `brew cleanup pkgconf`...
  Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
  Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
  Removing: /usr/local/Cellar/pkg-config/0.29.2_3... (12 files, 631.4KB)
  
  $ brew info pkg-config
  ==> pkgconf: stable 2.3.0 (bottled), HEAD
  Package compiler and linker metadata toolkit
  https://github.com/pkgconf/pkgconf
  Installed
  /usr/local/Cellar/pkgconf/2.3.0 (25 files, 327.9KB) *
    Poured from bottle on 2024-11-08 at 13:13:07
  From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/p/pkgconf.rb
  License: ISC
  ==> Options
  --HEAD
	  Install HEAD version

  $ jq '.source' /usr/local/opt/pkg-config/INSTALL_RECEIPT.json
  {
    "spec": "stable",
    "versions": {
      "stable": "2.3.0",
      "head": null,
      "version_scheme": 0
    },
    "path": "/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Aliases/pkg-config",
    "tap_git_head": "afeea04d898316ef1ea37d87b80de9c9829550d7",
    "tap": "homebrew/core"
  }
  ```

* `brew install <formula-with-pkg-config-dep>` will automatically install `pkgconf` and record it in receipt:
  ```console
  $ brew install pkg-config
  ==> Downloading https://ghcr.io/v2/homebrew/core/pkg-config/manifests/0.29.2_3
  Already downloaded: /Users/cho-m/Library/Caches/Homebrew/downloads/ac691fc7ab8ecffba32a837e7197101d271474a3a84cfddcc30c9fd6763ab3c6--pkg-config-0.29.2_3.bottle_manifest.json
  ==> Fetching pkg-config
  ==> Downloading https://ghcr.io/v2/homebrew/core/pkg-config/blobs/sha256:55f4268c3d6f320932ce736e8ad862a98b27048a30481f85c0679d8c655fb0f6
  Already downloaded: /Users/cho-m/Library/Caches/Homebrew/downloads/25f2808beac7c455c7101ea7fb7f9ae61f5019163147d7e822e30a76ceb11263--pkg-config--0.29.2_3.sequoia.bottle.tar.gz
  ==> Pouring pkg-config--0.29.2_3.sequoia.bottle.tar.gz
  🍺  /usr/local/Cellar/pkg-config/0.29.2_3: 12 files, 631.5KB
  ==> Running `brew cleanup pkg-config`...
  Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
  Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
  
  $ # update branch and revision
  
  $ brew install libpthread-stubs
  ==> Downloading https://ghcr.io/v2/homebrew/core/libpthread-stubs/manifests/0.5-1
  Already downloaded: /Users/cho-m/Library/Caches/Homebrew/downloads/daf0585704083a365a05b49b2a0078c97907e4a8b527f2a432b59a77805cf3f4--libpthread-stubs-0.5-1.bottle_manifest.json
  ==> Fetching dependencies for libpthread-stubs: pkgconf
  ==> Downloading https://ghcr.io/v2/homebrew/core/pkgconf/manifests/2.3.0
  Already downloaded: /Users/cho-m/Library/Caches/Homebrew/downloads/2f1a3719bf003aa9443a2e2e6af8e6441aa287088a24771231f191a9bcf36884--pkgconf-2.3.0.bottle_manifest.json
  ==> Fetching pkgconf
  ==> Downloading https://ghcr.io/v2/homebrew/core/pkgconf/blobs/sha256:9b44fe313d296fa10617d6b58ea1eab78bf217b6f00478ad98b01e0375475472
  Already downloaded: /Users/cho-m/Library/Caches/Homebrew/downloads/eed8a0a3462be8896edfce1cf6d726e4ebfd489cafecfb485f1b1b669a5d89af--pkgconf--2.3.0.sequoia.bottle.tar.gz
  ==> Fetching libpthread-stubs
  ==> Downloading https://ghcr.io/v2/homebrew/core/libpthread-stubs/blobs/sha256:303b8c21fc1b9322b6bd8f24e75a4e53a1c331d09b4f6271f75eba743d119819
  Already downloaded: /Users/cho-m/Library/Caches/Homebrew/downloads/623815c812e848149470976d73b58273ffb90c3873fe1ed3ec11f6d4a88c877f--libpthread-stubs--0.5.all.bottle.1.tar.gz
  ==> Installing dependencies for libpthread-stubs: pkgconf
  ==> Installing libpthread-stubs dependency: pkgconf
  ==> Downloading https://ghcr.io/v2/homebrew/core/pkgconf/manifests/2.3.0
  Already downloaded: /Users/cho-m/Library/Caches/Homebrew/downloads/2f1a3719bf003aa9443a2e2e6af8e6441aa287088a24771231f191a9bcf36884--pkgconf-2.3.0.bottle_manifest.json
  ==> Pouring pkgconf--2.3.0.sequoia.bottle.tar.gz
  🍺  /usr/local/Cellar/pkgconf/2.3.0: 25 files, 327.9KB
  ==> Installing libpthread-stubs
  ==> Pouring libpthread-stubs--0.5.all.bottle.1.tar.gz
  🍺  /usr/local/Cellar/libpthread-stubs/0.5: 6 files, 10.6KB
  ==> Running `brew cleanup libpthread-stubs`...
  Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
  Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).

  $ jq '.runtime_dependencies' /usr/local/Cellar/libpthread-stubs/0.5/INSTALL_RECEIPT.json
  [
    {
      "full_name": "pkgconf",
      "version": "2.3.0",
      "revision": 0,
      "pkg_version": "2.3.0",
      "declared_directly": false
    }
  ]
  ```

**Comment by MikeMcQuaid** [comment]
@cho-m sounds good, thanks!

**Review by carlocab — APPROVED** [review]

**Comment by cho-m** [comment]
Will rebase due to my Homebrew/brew PR which will probably cause syntax failure. Then can merge.

**Comment by deecewan** [comment]
is there any context as to why this was done? this seems to be the PR that deleted `pkg-config` formula, but it doesn't mention at all why it happened.

**Comment by carlocab** [comment]
See discussion at https://github.com/Homebrew/homebrew-core/pull/182175

**Comment by Bo98** [comment]
The change shouldn't have any notable effect for the vast majority of users (except one incident with GitHub Actions that was caused by GitHub themselves which they have since fixed). We added aliases so `brew install pkg-config` still works and `pkg-config` still exists as an executable.

If you face any regressions let us know.

**Pull request merged**